proposals/0534-metro-package-exports-support.md

```diff
@@ -0,0 +1,455 @@
+
+- **P0 - Core features**: The robust subset of features that we must support in the spec for a minimum implementation of `"exports"`.
+    - This work will be scheduled ahead of P1 features unless it makes sense not to during implementation.
+    - App developers may wish to opt in to experimental `"exports"` support upon completion of these features.
+- **P1 - Not targeted in first iteration**:
+    - Attached to features which are more challenging to implement, or that were launched in later revisions to the Node.js spec.
+    - These may still be done, otherwise will be fast-follow.
+    - This work may be up for grabs by interested external contributors.
+
+**All features discussed are P0 unless otherwise stated.**
+
+## Detailed design
+
+We intend to support all features of `"exports"` up to Node 18.10. For now, we are aware of but are not committing to support for [`"imports"`](https://nodejs.org/docs/latest-v18.x/api/packages.html#imports).
+
+We are beginning in a state where `"exports"` is in general use and we cannot guarantee that references to these packages within existing React Native projects are compliant with the `"exports"` spec. Therefore Metro needs to carefully design around this.
+
+The following features and considerations are detailed in this section:
+
+- [Package encapsulation](#package-encapsulation)
+- [Subpath exports](#subpath-exports)
+- [Subpath patterns](#subpath-patterns)
+- [Exact path specifiers](#exact-path-specifiers)
+- [Conditional exports](#conditional-exports)
+- [Conditional exports: `"import"` and `"require"`](#conditional-exports-import-and-require)
+- [Conditional exports: Community definitions and `"browser"`](#conditional-exports-community-definitions-and-browser)
+- [Conditional exports: User conditions](#conditional-exports-user-conditions)
+- [Opting into strict `"exports"` handling](#opting-into-strict-exports-handling)
+
+The following features will be implemented without any anticipated behaviour differences or special notes (links go to Node.js spec):
+
+- [Exports sugar](https://nodejs.org/docs/latest-v18.x/api/packages.html#exports-sugar)
+- [Self-referencing a package using its name](https://nodejs.org/docs/latest-v18.x/api/packages.html#self-referencing-a-package-using-its-name)
+
+### Package encapsulation
+
+This is the primary breaking change introduced by `"exports"`.
+
+> When the `"exports"` field is defined, all subpaths of the package are encapsulated and no longer available to importers.\
+– https://nodejs.org/docs/latest-v18.x/api/packages.html#main-entry-point-export
+
+```json
+"exports": {
+  // Paths outside of this list are inaccessible to importers
+  ".": "./lib/index.js",
+  "./package.json": "./package.json"
+}
+```
+
+**Proposed**:
+
+- By default, Metro will ignore this constraint. We will log a warning when an inaccessible import is accessed.
```

Comment:
One concern I have with these logs is that the user is not the owner of the library, and can't change the code without the library author (`unless using patch-package, but that shouldn't be a default`).

As long as the log mentions pursuing the library authors to change it, I think this is great. If that's not clear, that it's not an issue with the project itself, I think it only confuses users.

Reply:
Yes we plan to make both sides of the warning clear. Provisional text for the warning is [here](https://github.com/react-native-community/discussions-and-proposals/pull/534/files#diff-40781647168e437e711b426935003e99ce4bddb4c1a5f1ebd61b0218d2b45670R399).